src/adc.rs

```diff
@@ -0,0 +1,562 @@
+    gpioe::PE7<Analog> => 13_u8,
+    gpioe::PE10<Analog> => 14_u8,
+    gpioe::PE11<Analog> => 15_u8,
+    gpioe::PE12<Analog> => 16_u8,
+    // Shared channels (i.e. ADC34_INx)
+    gpioe::PE8<Analog> => 6_u8,
+    gpiod::PD10<Analog> => 7_u8,
+    gpiod::PD11<Analog> => 8_u8,
+    gpiod::PD12<Analog> => 9_u8,
+    gpiod::PD13<Analog> => 10_u8,
+    gpiod::PD14<Analog> => 11_u8,
+);
+
+// # ADC4 Pin/Channel mapping
+// ## f303
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+adc_pins!(ADC4,
+    gpioe::PE14<Analog> => 1_u8,
+    gpioe::PE15<Analog> => 2_u8,
+    gpiob::PB12<Analog> => 3_u8,
+    gpiob::PB14<Analog> => 4_u8,
+    gpiob::PB15<Analog> => 5_u8,
+    gpiob::PB8<Analog> => 12_u8,
+    gpiob::PB9<Analog> => 13_u8,
+    // Shared channels (i.e. ADC34_INx)
+    gpioe::PE8<Analog> => 6_u8,
+    gpiod::PD10<Analog> => 7_u8,
+    gpiod::PD11<Analog> => 8_u8,
+    gpiod::PD12<Analog> => 9_u8,
+    gpiod::PD13<Analog> => 10_u8,
+    gpiod::PD14<Analog> => 11_u8,
+);
+
+/// Abstract implementation of ADC functionality
+///
+/// Do not use directly. See adc12_hal for a applicable Macro.
+/// TODO: Extend/generalize beyond f303
+macro_rules! adc_hal {
+    ($(
+            $ADC:ident: ($adcx:ident, $ADC_COMMON:ident),
+    )+) => {
+        $(
+            impl Adc<$ADC> {
+
+                /// Init a new ADC
+                ///
+                /// Enables the clock, performs a calibration and enables the ADC
+                pub fn $adcx(
+                    rb: $ADC,
+                    adc_common : &mut $ADC_COMMON,
+                    ahb: &mut AHB,
+                    ckmode: CKMODE,
+                    clocks: Clocks,
+                ) -> Option<Self> {
+                    let mut this_adc = Self {
+                        rb,
+                        clocks,
+                        ckmode,
+                        operation_mode: None,
+                    };
+                    if !(this_adc.clocks_welldefined(clocks)) {
+                        return None;
+                    }
+                    if !(this_adc.enable_clock(ahb, adc_common)){
+                        return None;
+                    }
+                    this_adc.set_align(Align::default());
+                    this_adc.calibrate();
+                    // ADEN bit cannot be set during ADCAL=1
+                    // and 4 ADC clock cycle after the ADCAL
+                    // bit is cleared by hardware
+                    this_adc.wait_adc_clk_cycles(4);
+                    this_adc.enable();
+                    Some(this_adc)
+                }
+
+                /// Software can use CKMODE::SYNCDIV1 only if
+                /// hclk and sysclk are the same. (see reference manual 15.3.3)
+                fn clocks_welldefined(&self, clocks: Clocks) -> bool {
+                    !(self.ckmode == CKMODE::SYNCDIV1 && !(clocks.hclk().0 == clocks.sysclk().0))
```

Comment:
I really had trouble understanding this condition. I think this would really benefit from being a bit more verbose, like:
```rust
if self.ckmode == CHKMODE::SYNCDIV1 {
    clocks.hclk().0 == clocks.sysclk().0
} else {
    true
}
```
Or, if it absolutely needs to be one line, at least get rid of the double negation:
```rust
self.ckmode != CKMODE::SYNCDIV1 || clocks.hclk().0 == clocks.sysclk().0
```